File: workers-api/src/services/progress-service.ts
import { eq, and, sql, count } from "drizzle-orm";
import type { Database } from "../db/client";
import { questions, userProgress, bookmarks } from "../db/schema";

export interface ProgressSummary {
  total: number;
  attempted: number;
  correct: number;
  incorrect: number;
  bookmarked: number;
}

export async function getProgressSummary(
  db: Database,
  examId: number,
  userId: number
): Promise<ProgressSummary> {
  const [totalResult] = await db
    .select({ count: count() })
    .from(questions)
    .where(eq(questions.examId, examId));

  const progressRows = await db
    .select({
      isCorrect: userProgress.isCorrect,
      cnt: count(),
    })
    .from(userProgress)
    .innerJoin(questions, eq(userProgress.questionId, questions.id))
    .where(and(eq(userProgress.userId, userId), eq(questions.examId, examId)))
    .groupBy(userProgress.isCorrect);

  let correct = 0;
  let incorrect = 0;
  for (const row of progressRows) {
    if (row.isCorrect) correct = row.cnt;
    else incorrect = row.cnt;
  }

  const [bookmarkResult] = await db
    .select({ count: count() })
    .from(bookmarks)
    .innerJoin(questions, eq(bookmarks.questionId, questions.id))
    .where(and(eq(bookmarks.userId, userId), eq(questions.examId, examId)));

  return {
    total: totalResult.count,
    attempted: correct + incorrect,
    correct,
    incorrect,
    bookmarked: bookmarkResult.count,
  };
}

export interface ProgressDetailItem {
  question_id: number;
  external_id: string;
  is_correct: boolean;
  selected_option_ids: number[];
}

export async function getProgressDetail(
  db: Database,
  examId: number,
  userId: number
): Promise<{ summary: ProgressSummary; items: ProgressDetailItem[] }> {
  const summary = await getProgressSummary(db, examId, userId);

  const rows = await db
    .select({
      questionId: userProgress.questionId,
      externalId: questions.externalId,
      isCorrect: userProgress.isCorrect,
      selectedOptionIds: userProgress.selectedOptionIds,
    })
    .from(userProgress)
    .innerJoin(questions, eq(userProgress.questionId, questions.id))
    .where(and(eq(userProgress.userId, userId), eq(questions.examId, examId)));

  const items = rows.map((r) => ({
    question_id: r.questionId,
    external_id: r.externalId,
    is_correct: r.isCorrect,
    selected_option_ids: JSON.parse(r.selectedOptionIds) as number[],
  }));

  return { summary, items };
}

export async function resetProgress(
  db: Database,
  examId: number,
  userId: number
): Promise<void> {
  const questionIds = db
    .select({ id: questions.id })
    .from(questions)
    .where(eq(questions.examId, examId));

  await db
    .delete(userProgress)
    .where(
      and(
        eq(userProgress.userId, userId),
        sql`${userProgress.questionId} IN (${questionIds})`
      )
    );
}

File: backend/app/services/progress_service.py
import json

from sqlalchemy import delete, func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models import Bookmark, Question, UserProgress
from app.schemas.exam import ProgressSummary
from app.schemas.progress import ProgressDetailItem, ProgressDetailResponse


async def get_progress_summary(db: AsyncSession, exam_id: int, user_id: int) -> ProgressSummary:
    # Total questions in exam
    total_result = await db.execute(
        select(func.count()).select_from(Question).where(Question.exam_id == exam_id)
    )
    total = total_result.scalar() or 0

    # Get question IDs for this exam
    q_ids_result = await db.execute(
        select(Question.id).where(Question.exam_id == exam_id)
    )
    question_ids = [row[0] for row in q_ids_result.all()]

    if not question_ids:
        return ProgressSummary(total=total)

    # Attempted
    progress_result = await db.execute(
        select(UserProgress)
        .where(UserProgress.user_id == user_id, UserProgress.question_id.in_(question_ids))
    )
    progress_items = progress_result.scalars().all()
    correct = sum(1 for p in progress_items if p.is_correct)
    incorrect = len(progress_items) - correct

    # Bookmarked
    bm_result = await db.execute(
        select(func.count())
        .select_from(Bookmark)
        .where(Bookmark.user_id == user_id, Bookmark.question_id.in_(question_ids))
    )
    bookmarked = bm_result.scalar() or 0

    return ProgressSummary(
        total=total,
        attempted=len(progress_items),
        correct=correct,
        incorrect=incorrect,
        bookmarked=bookmarked,
    )


async def get_progress_detail(db: AsyncSession, exam_id: int, user_id: int) -> ProgressDetailResponse:
    summary = await get_progress_summary(db, exam_id, user_id)

    q_ids_result = await db.execute(
        select(Question.id).where(Question.exam_id == exam_id)
    )
    question_ids = [row[0] for row in q_ids_result.all()]

    progress_result = await db.execute(
        select(UserProgress, Question.external_id)
        .join(Question, UserProgress.question_id == Question.id)
        .where(UserProgress.user_id == user_id, UserProgress.question_id.in_(question_ids))
    )

    items = [
        ProgressDetailItem(
            question_id=p.question_id,
            external_id=ext_id,
            is_correct=p.is_correct,
            selected_option_ids=json.loads(p.selected_option_ids),
        )
        for p, ext_id in progress_result.all()
    ]

    return ProgressDetailResponse(summary=summary, items=items)


async def reset_progress(db: AsyncSession, exam_id: int, user_id: int) -> None:
    q_ids_result = await db.execute(
        select(Question.id).where(Question.exam_id == exam_id)
    )
    question_ids = [row[0] for row in q_ids_result.all()]

    if question_ids:
        await db.execute(
            delete(UserProgress).where(
                UserProgress.user_id == user_id, UserProgress.question_id.in_(question_ids)
            )
        )
        await db.commit()
